### PR TITLE
fix a dropdown button menu position bug

### LINKED ISSUE #89029: DropdownButton overlay position issue with multiple items
Dropdown overlay is wrongly positioned when there are multiple items in [items](https://api.flutter.dev/flutter/material/DropdownButton/items.html) property of DropdownButton widget along with [menuMaxHeight](https://api.flutter.dev/flutter/material/DropdownButton/menuMaxHeight.html) property

## Steps to Reproduce

1. Navigate to official flutter [DropdownButton](https://api.flutter.dev/flutter/material/DropdownButton-class.html) widget documentation
2. Update line `55` by adding more items to DropdownButton `items` property:
```dart
items: <String>['One', 'Two', 'Free', 'Four', 'Five', 'Six', 'Seven', 'Eight', 'Nine', 'Ten', 'Eleven', 'Twelve']
```
3. Add `menuMaxHeight` property to DropdownButton Widget and set its value:
```dart
menuMaxHeight: 100,
```
4. Run code 

**Expected results:** 
image

**Actual results:**
image

### PR DESCRIPTION
**Pull request by xu-baolin**
Fixes #89029 

This is an enhanced patch of #76493

### CODE REVIEW

File: packages/flutter/lib/src/material/dropdown.dart [hunk 1]
```diff
@@ -543,14 +546,21 @@ class _DropdownRoute<T> extends PopupRoute<_DropdownRouteResult<T>> {
     // than the button height and the button is at the very bottom or top of the
     // screen, the menu will be aligned with the bottom or top of the button
     // respectively.
-    if (menuTop < topLimit)
+    if (menuTop < topLimit) {
       menuTop = math.min(buttonTop, topLimit);
+      menuBottom = menuTop + menuHeight;
```
Comment (xu-baolin) [inline packages/flutter/lib/src/material/dropdown.dart:551]: This is a bug and I add an assertion to check this at line 580

**Comment by xu-baolin** [comment]
Ready for review now.

File: packages/flutter/lib/src/material/dropdown.dart [hunk 2]
```diff
@@ -516,7 +516,10 @@ class _DropdownRoute<T> extends PopupRoute<_DropdownRouteResult<T>> {
   // selected item is aligned with the button's vertical center, as far as
   // that's possible given availableHeight.
   _MenuLimits getMenuLimits(Rect buttonRect, double availableHeight, int index) {
-    final double maxMenuHeight = availableHeight - 2.0 * _kMenuItemHeight;
+    double maxMenuHeight = availableHeight - 2.0 * _kMenuItemHeight;
```
Comment (Piinks) [inline packages/flutter/lib/src/material/dropdown.dart:519]: `maxMenuHeight` v `menuMaxHeight` is a bit confusing.
What if we made this one `computedMaxHeight`? Or something like that?

File: packages/flutter/lib/src/material/dropdown.dart [hunk 3]
```diff
@@ -567,6 +577,7 @@ class _DropdownRoute<T> extends PopupRoute<_DropdownRouteResult<T>> {
       scrollOffset = math.min(scrollOffset, preferredMenuHeight - menuHeight);
     }
 
+    assert(menuHeight == menuBottom - menuTop);
```
Comment (Piinks) [inline packages/flutter/lib/src/material/dropdown.dart:580]: Can you add a test that triggers this assertion?

Comment (xu-baolin) [reply packages/flutter/lib/src/material/dropdown.dart:580]: If we only add this assertion to the master branch, there will be many tests that trigger this assertion.
 Fortunately, no one has used `_MenuLimits.menuBottom` before, so there are no functional issues.

Comment (xu-baolin) [reply packages/flutter/lib/src/material/dropdown.dart:581]: This is just to check that the metrics calculated by this function are correct. 
I didn't think of a good way to test and trigger it.

**Review by Piinks — APPROVED** [review]
LGTM! This will need to be rebased after https://github.com/flutter/flutter/pull/89861 lands, it is the cause of the test failures here

**Pull request merged**